[CI] SizeBlockingQueueTests.testQueueSize fails on 6.x

I am seeing this failure every once in a while:

```
Suite: org.elasticsearch.common.util.concurrent.SizeBlockingQueueTests
  1> [2018-02-08T18:01:44,333][INFO ][o.e.c.u.c.SizeBlockingQueueTests] [testQueueSize]: before test
  1> [2018-02-08T18:01:44,355][INFO ][o.e.c.u.c.SizeBlockingQueueTests] [testQueueSize]: after test
  2> REPRODUCE WITH: ./gradlew :server:test -Dtests.seed=69113FDE6E332977 -Dtests.class=org.elasticsearch.common.util.concurrent.SizeBlockingQueueTests -Dtests.method="testQueueSize" -Dtests.security.manager=true -Dtests.jvms=3 -Dtests.locale=id-ID -Dtests.timezone=Africa/Conakry
FAILURE 0.04s J2 | SizeBlockingQueueTests.testQueueSize <<< FAILURES!
   > Throwable #1: java.lang.AssertionError: 
   > Expected: <20>
   >      but: was <0>
   > 	at __randomizedtesting.SeedInfo.seed([69113FDE6E332977:1884A5ABD4B7A214]:0)
   > 	at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:20)
   > 	at org.elasticsearch.common.util.concurrent.SizeBlockingQueueTests.testQueueSize(SizeBlockingQueueTests.java:91)
   > 	at java.lang.Thread.run(Thread.java:748)
  2> NOTE: leaving temporary files on disk at: /home/hinmanm/es/elasticsearch/server/build/testrun/test/J2/temp/org.elasticsearch.common.util.concurrent.SizeBlockingQueueTests_69113FDE6E332977-001
  2> NOTE: test params are: codec=Asserting(Lucene70): {}, docValues:{}, maxPointsInLeafNode=374, maxMBSortInHeap=7.113482659052134, sim=RandomSimilarity(queryNorm=true): {}, locale=id-ID, timezone=Africa/Conakry
  2> NOTE: Linux 4.14.16-300.fc27.x86_64 amd64/Oracle Corporation 1.8.0_151 (64-bit)/cpus=8,threads=1,free=330397240,total=532152320
```

When running *all* tests (I haven't been able to reproduce it running as a single test)

Related to #28557

--- a/server/src/test/java/org/elasticsearch/common/util/concurrent/SizeBlockingQueueTests.java
+++ b/server/src/test/java/org/elasticsearch/common/util/concurrent/SizeBlockingQueueTests.java
@@ -24,9 +24,7 @@ import org.elasticsearch.test.ESTestCase;
 import java.util.concurrent.ArrayBlockingQueue;
 import java.util.concurrent.BlockingQueue;
 import java.util.concurrent.BrokenBarrierException;
-import java.util.concurrent.CountDownLatch;
 import java.util.concurrent.CyclicBarrier;
-import java.util.concurrent.atomic.AtomicBoolean;
 import java.util.concurrent.atomic.AtomicInteger;
 
 import static org.hamcrest.Matchers.equalTo;
@@ -48,38 +46,12 @@ public class SizeBlockingQueueTests extends ESTestCase {
             sizeBlockingQueue.offer(i);
         }
 
-        final CountDownLatch latch = new CountDownLatch(1);
-        final CyclicBarrier barrier = new CyclicBarrier(2);
 
-        final AtomicInteger maxSize = new AtomicInteger();
-
-        // this thread will repeatedly poll the size of the queue keeping track of the maximum size that it sees
         final int iterations = 1 << 16;
-        final Thread queueSizeThread = new Thread(() -> {
-            try {
-                latch.await();
-            } catch (final InterruptedException e) {
-                throw new RuntimeException(e);
-            }
-            for (int i = 0; i < iterations; i++) {
-                try {
-                    // synchronize each iteration of checking the size with each iteration of offering, each iteration is a race
-                    barrier.await();
-                } catch (final BrokenBarrierException | InterruptedException e) {
-                    throw new RuntimeException(e);
-                }
-                maxSize.set(Math.max(maxSize.get(), sizeBlockingQueue.size()));
-            }
-        });
-        queueSizeThread.start();
+        final CyclicBarrier barrier = new CyclicBarrier(2);
 
         // this thread will try to offer items to the queue while the queue size thread is polling the size
         final Thread queueOfferThread = new Thread(() -> {
-            try {
-                latch.await();
-            } catch (final InterruptedException e) {
-                throw new RuntimeException(e);
-            }
             for (int i = 0; i < iterations; i++) {
                 try {
                     // synchronize each iteration of checking the size with each iteration of offering, each iteration is a race
@@ -92,8 +64,20 @@ public class SizeBlockingQueueTests extends ESTestCase {
         });
         queueOfferThread.start();
 
-        // synchronize the start of the two threads
-        latch.countDown();
+        // this thread will repeatedly poll the size of the queue keeping track of the maximum size that it sees
+        final AtomicInteger maxSize = new AtomicInteger();
+        final Thread queueSizeThread = new Thread(() -> {
+            for (int i = 0; i < iterations; i++) {
+                try {
+                    // synchronize each iteration of checking the size with each iteration of offering, each iteration is a race
+                    barrier.await();
+                } catch (final BrokenBarrierException | InterruptedException e) {
+                    throw new RuntimeException(e);
+                }
+                maxSize.set(Math.max(maxSize.get(), sizeBlockingQueue.size()));
+            }
+        });
+        queueSizeThread.start();
 
         // wait for the threads to finish
         queueOfferThread.join();

--- a/server/src/test/java/org/elasticsearch/common/util/concurrent/SizeBlockingQueueTests.java
+++ b/server/src/test/java/org/elasticsearch/common/util/concurrent/SizeBlockingQueueTests.java
@@ -23,7 +23,9 @@ import org.elasticsearch.test.ESTestCase;
 
 import java.util.concurrent.ArrayBlockingQueue;
 import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.BrokenBarrierException;
 import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.CyclicBarrier;
 import java.util.concurrent.atomic.AtomicBoolean;
 import java.util.concurrent.atomic.AtomicInteger;
 
@@ -47,18 +49,26 @@ public class SizeBlockingQueueTests extends ESTestCase {
         }
 
         final CountDownLatch latch = new CountDownLatch(1);
+        final CyclicBarrier barrier = new CyclicBarrier(2);
 
         final AtomicBoolean spin = new AtomicBoolean(true);
         final AtomicInteger maxSize = new AtomicInteger();
 
         // this thread will repeatedly poll the size of the queue keeping track of the maximum size that it sees
+        final int iterations = 1 << 16;
         final Thread queueSizeThread = new Thread(() -> {
             try {
                 latch.await();
             } catch (final InterruptedException e) {
                 throw new RuntimeException(e);
             }
-            while (spin.get()) {
+            for (int i = 0; i < iterations; i++) {
+                try {
+                    // synchronize each iteration of checking the size with each iteration of offering, each iteration is a race
+                    barrier.await();
+                } catch (final BrokenBarrierException | InterruptedException e) {
+                    throw new RuntimeException(e);
+                }
                 maxSize.set(Math.max(maxSize.get(), sizeBlockingQueue.size()));
             }
         });
@@ -71,7 +81,13 @@ public class SizeBlockingQueueTests extends ESTestCase {
             } catch (final InterruptedException e) {
                 throw new RuntimeException(e);
             }
-            for (int i = 0; i < 4096; i++) {
+            for (int i = 0; i < iterations; i++) {
+                try {
+                    // synchronize each iteration of checking the size with each iteration of offering, each iteration is a race
+                    barrier.await();
+                } catch (final BrokenBarrierException | InterruptedException e) {
+                    throw new RuntimeException(e);
+                }
                 sizeBlockingQueue.offer(capacity + i);
             }
         });

--- a/server/src/test/java/org/elasticsearch/common/util/concurrent/SizeBlockingQueueTests.java
+++ b/server/src/test/java/org/elasticsearch/common/util/concurrent/SizeBlockingQueueTests.java
@@ -51,7 +51,6 @@ public class SizeBlockingQueueTests extends ESTestCase {
         final CountDownLatch latch = new CountDownLatch(1);
         final CyclicBarrier barrier = new CyclicBarrier(2);
 
-        final AtomicBoolean spin = new AtomicBoolean(true);
         final AtomicInteger maxSize = new AtomicInteger();
 
         // this thread will repeatedly poll the size of the queue keeping track of the maximum size that it sees
@@ -96,11 +95,8 @@ public class SizeBlockingQueueTests extends ESTestCase {
         // synchronize the start of the two threads
         latch.countDown();
 
-        // wait for the offering thread to finish
+        // wait for the threads to finish
         queueOfferThread.join();
-
-        // stop the queue size thread
-        spin.set(false);
         queueSizeThread.join();
 
         // the maximum size of the queue should be equal to the capacity

